flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/TableUpdater.java

```diff
@@ -0,0 +1,205 @@
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.UpdatePartitionSpec;
+import org.apache.iceberg.UpdateSchema;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.SupportsNamespaces;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Internal
+class TableUpdater {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TableUpdater.class);
+  private final TableDataCache cache;
+  private final Catalog catalog;
+
+  TableUpdater(TableDataCache cache, Catalog catalog) {
+    this.cache = cache;
+    this.catalog = catalog;
+  }
+
+  /**
+   * Creates or updates a table to make sure that the given branch, schema, spec exists.
+   *
+   * @return a {@link Tuple3} of the new {@link Schema}, the status of the schema compared to the
+   *     requested one, and the new {@link PartitionSpec#specId()}.
+   */
+  Tuple3<Schema, CompareSchemasVisitor.Result, PartitionSpec> update(
+      TableIdentifier tableIdentifier, String branch, Schema schema, PartitionSpec spec) {
+    findOrCreateTable(tableIdentifier, schema, spec);
+    findOrCreateBranch(tableIdentifier, branch);
+    Tuple2<Schema, CompareSchemasVisitor.Result> newSchema =
+        findOrCreateSchema(tableIdentifier, schema);
+    PartitionSpec newSpec = findOrCreateSpec(tableIdentifier, spec);
+    return Tuple3.of(newSchema.f0, newSchema.f1, newSpec);
+  }
+
+  private void findOrCreateTable(TableIdentifier identifier, Schema schema, PartitionSpec spec) {
+    Tuple2<Boolean, Exception> exists = cache.exists(identifier);
+    if (Boolean.FALSE.equals(exists.f0)) {
+      if (exists.f1 instanceof NoSuchNamespaceException) {
+        SupportsNamespaces catalogWithNameSpace = (SupportsNamespaces) catalog;
+        LOG.info("Namespace {} not found during table search. Creating namespace", identifier);
+        try {
+          catalogWithNameSpace.createNamespace(identifier.namespace());
+        } catch (AlreadyExistsException e) {
+          LOG.debug("Namespace {} created concurrently", identifier.namespace(), e);
+        }
+
+        createTable(identifier, schema, spec);
+      } else {
+        LOG.info("Table {} not found during table search. Creating table.", identifier);
+        createTable(identifier, schema, spec);
+      }
+    }
+  }
+
+  private void createTable(TableIdentifier identifier, Schema schema, PartitionSpec spec) {
+    try {
+      Table table = catalog.createTable(identifier, schema, spec);
+      cache.update(identifier, table);
+    } catch (AlreadyExistsException e) {
+      LOG.info("Table {} created concurrently. Skipping creation.", identifier, e);
```

Comment:
Do we need these logs on info level? If we have many records, then we might end up flooding the logs?
Could you please check where we can use debug level without losing traceability?

Reply:
Yes, it absolutely makes sense to have this at debug level. This will get printed frequently on concurrent new table creation. It might be confusing to users although it is nothing concerning.